scripts: blobs: add license-click-through property

yangbolu1991 opened the pull request:
Added a license-click-through property. Some downloading may need to accept license to continue.

And an example using it is https://github.com/zephyrproject-rtos/hal_nxp/pull/515

Thanks.

yangbolu1991 committed:
scripts: blobs: add click-through property

Added a click-through property. Some downloading may need to
accept license to continue.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

carlescufi reviewed (dismissed):
LGTM, thank you!

carlescufi commented:
@dleach02 @kartben should we update any documentation with this change? Not strictly necessary IMHO, but could be good.

pdgendt commented on scripts/west_commands/blobs.py:
```diff
@@ -151,6 +153,23 @@ def fetch(self, args):
             self.inf('Fetching blob {module}: {abspath}'.format(**blob))
+
+            if blob['license-click-through']:
+                user_input = input("For this blob, need to read and accept license to continue."
+                                   "Read it? [y/n]: ")
```
Splitting this over multiple lines will still concatenate it without spacing. Also not sure about the text itself.

```suggestion
                user_input = input("For this blob, need to read and accept license to continue. "
                                   "Read it? [y/n]: ")
```

yangbolu1991 replied:
Thanks. Will change as this.

pdgendt commented on scripts/west_commands/blobs.py:
```diff
@@ -151,6 +153,23 @@ def fetch(self, args):
+
+                user_input = input("Accept license to continue? [y/n]: ")
+                if user_input.upper() != "Y":
+                    self.inf('Warn: skip fetching this blob.')
```
Printing a warning with `inf`?

```suggestion
                    self.wrn('Skipped fetching this blob.')
```

yangbolu1991 replied:
Thanks. Will use it.

pdgendt commented on scripts/west_commands/blobs.py:
```diff
@@ -55,6 +55,8 @@ def do_add_parser(self, parser_adder):
             - version: version string
             - license_path: path to the license file for the blob
+            - license-abspath: absolute path to the license file for the blob
+            - license-click-through: need license click-through or not
```
Is `click-through` clear on what the purpose is?

yangbolu1991 replied:
Yes. I'm okay to use click-through.
Thanks.

pdgendt replied:
That's not what I meant, I find the wording of `click-through` a bit odd, maybe `license-agreement-required` or something is a better fit?

dleach02 replied:
@pdgendt clickthrough or clickwrap are known licensing terms where the underlying license requires presenting and obtaining consent to the terms before downloading. It is better to stick with known terminology. 

https://en.wikipedia.org/wiki/Clickwrap

tejlmand commented on scripts/west_commands/blobs.py:
```diff
@@ -151,6 +153,23 @@ def fetch(self, args):
+
+            if blob['license-click-through']:
+                user_input = input("For this blob, need to read and accept license to continue."
+                                   "Read it? [y/n]: ")
+                if user_input.upper() != "Y":
+                    self.inf('Warn: skip fetching this blob.')
```
should we only test for `Y` ?
Or should we actually test specifically `Y` and `N` and request user to use one of those, so that typing `yes` will re-request the user to type `y/n`.

Right now, if a user types `yes<enter>`, then the west command will continue as if user typed `N`.

yangbolu1991 replied:
Will request user to input y or n, and check strictly.
Thanks.

tejlmand reviewed (changes requested):
When requesting user input, then I think it's important that we only continue if user actually types one of the approved input values, and not accept all other values besides `Y` as an `N`.

yangbolu1991 commented:
Updated to v2. Changes include,
- Converted to `click-through`
- Converted to use `self.wrn`
- Requested user to input y or n, and check strictly.

Thanks.

carlescufi commented on scripts/west_commands/blobs.py:
```diff
@@ -151,6 +153,33 @@ def fetch(self, args):
+
+            if blob['click-through']:
+                while True:
+                    user_input = input("For this blob, need to read and accept "
+                                       "license to continue. Read it?\n"
+                                       "Please type 'y' or 'n' and press enter to confirm: ")
+                    if user_input.upper() == "Y" or user_input.upper() == "N":
+                        break
```
Blobs are never fetched in CI though. And we can always add a `-y` option later.

pdgendt replied:
Downstream CI could (I know we do)

pdgendt reviewed (approved):
I wouldn't mind the "auto-accept" argument to be able to keep using it in an automated way.
Can be done in a follow-up PR.

yangbolu1991 commented:
> I wouldn't mind the "auto-accept" argument to be able to keep using it in an automated way. Can be done in a follow-up PR.

I created a PR. Please check if it address the problem:)
https://github.com/zephyrproject-rtos/zephyr/pull/90365

Thanks.
